Async snippet 2

jaredhirsch opened the pull request:
@shane-tomlinson mind taking a look? I could use some feedback.
#### Summary

This is a more well-rounded successor to the async snippet prototype I started last week (#3571). Seeking feedback.
#### Next Steps

If this looks good, I think the next step would be to create some RP integration docs (blog post and/or MDN page), and _maybe_ create a separate local testing example RP. Does this seem reasonable? Thoughts?
#### Background

We're in a bit of an awkward position relative to your typical third-party provider because we are shimming browser APIs, so our async snippet really does want to set global state. So, an async Persona snippet needs two parts: (1) something in the head that stubs out nav.id.\* before the script has loaded and sets up queues to collect function calls, and (2) the usual async snippet which builds a script tag, appends to the page, then unqueues waiting function calls when the script is ready.

This two-part thing seems complicated for a typical RP to have to integrate. I'm not sure.

The alternative I can think of is to put both parts of the snippet in the head, and wrap part (2) in a window.onload handler. This would also give additional flexibility, since the listener could be configured to listen for custom events--if the RP wanted to load some of their content scripts, and then fire nav.id.watch, for instance.

jaredhirsch committed:
example rp with first hacky async snippet in head, works. now to move to body.

jaredhirsch committed:
async snippet really async now. next, turn into a clean SDK.

jaredhirsch committed:
Cleanup the async snippet

shane-tomlinson commented:
Dude, this is fun.

After _a lot_ of hand tweaking, I was able to get it down to 740 bytes after being run through uglifyjs

https://gist.github.com/shane-tomlinson/5884391

shane-tomlinson commented:
The code in that gist works in IE8 and FF. Haven't tested others yet.

shane-tomlinson commented:
@6a68 - My opinion is that, if this were to be released, both halves should be combined into a single snippet that can be included. RPs would include the snippet whenever they are ready to start using navigator.id functions. We want to make it as simple as possible for RPs to be able to get the performance wins out of an async load.

Does this get us any closer to the goal of allowing RPs to self host include.js? 

I believe it does, I am trying to imagine how a message proxy would work using a pre-defined API like what is occurring here.

jaredhirsch commented:
@shane-tomlinson Thanks for having a look! Nicely done with that gist, great stuff :-D

I was thinking the same thing--splitting the snippet is a lot of complexity to ask RPs to manage. The problem is that, in order to not ask RPs to use a different async API, we have to stub out nav.id.\* globals with a queue really early in the page load, so we do need to be in the head--on the other hand, we could put the embed code in a tiny locally-hosted file, which takes the sting out a bit. We can defer the script tag loading until onload, to absolutely, definitely not affect time-to-content. I can't think of a simpler way...

On the plus side, waiting to load include.js until 'onload' fires means we could also wait until some RP-specified custom event fires, like 'onContentLoaded' or something. We can make this easy-to-specify by having room in the snippet for a custom event variable, something like what disqus does:

``` html
<!-- disqus snippet as example of specifying config vars -->
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = '<example>'; // Required - Replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
```

jaredhirsch commented:
@shane-tomlinson About self-hosting: arguably the embed snippet is the self-hostable part. Especially if we allow async loading (maybe another config option they can set), I think it's reasonable to tell RPs not to self-host, maybe?

Again, it's odd doing third-party stuff in a browser shim context, because, someday, the APIs will just be in the page, and delayed-loading semantics won't make a ton of sense. But, I think this would help adoption for now.

seanmonstar commented:
It'd definitely be awesome to not load the entire include.js if the async
shim detects native support already exists.

On Fri, Jun 28, 2013 at 8:27 AM, Jared Hirsch notifications@github.comwrote:

> @shane-tomlinson https://github.com/shane-tomlinson About self-hosting:
> arguably the embed snippet is the self-hostable part. Especially if we
> allow async loading (maybe another config option they can set), I think
> it's reasonable to tell RPs not to self-host, maybe?
> 
> Again, it's odd doing third-party stuff in a browser shim context,
> because, someday, the APIs will just be in the page, and delayed-loading
> semantics won't make a ton of sense. But, I think this would help adoption
> for now.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mozilla/browserid/pull/3590#issuecomment-20195019
> .

jaredhirsch committed:
Unify snippet in head, defer till DOM ready.

    * Adding @shane-tomlinson's awesome work from
      https://gist.github.com/shane-tomlinson/5884391
    * Instead of a two-part snippet, move the whole thing into the doc head
      and wait to append the async script tag until the DOM is ready.

jaredhirsch commented:
@shane-tomlinson @seanmonstar Updated. Now we put the whole embed blob in the doc head, wait until DOMContentLoaded (or IE's doc.onreadystatechange == 'complete'), and then do the async loading-and-dequeueing routine.

This also includes shane's fantastic changes from that gist he linked above, awesomesauce

jaredhirsch commented on example/rp/index.html:
```diff
@@ -47,6 +47,73 @@
+        // _qd will no longer be available.
+        if (nav.id._qd) {
+          // bail till ready, back off slowly with no timeout because EDGE/2G
+          return setTimeout(onParsed, parseWait+=50);
```
I also changed the polling logic to slowly back off forever, instead of giving up after waiting 10 seconds for parsing.

I'm hoping this will avoid a really spooky class of bugs, like the browser backgrounded the tab but it still runs setTimeouts periodically, or batches a thousand of them and runs them when the tab is foregrounded, etc.

jaredhirsch replied:
I'll fix that comment to not refer to network slowness, derp.

jaredhirsch committed:
Improve compressibility by assigning names.

    Reduces snippet size from 916 to 853 bytes with no other code changes \o/

jaredhirsch committed:
remove misleading comment

jaredhirsch commented:
just screwing around, I found that assigning reused strings like 'DOMContentLoaded' to variables reduced the size a bit.

@shane-tomlinson how do you test this in IE8 locally? maybe I need to finally get VM networking ironed out on my lappy.

chilts commented:
Not sure if this is helpful, but I have seen the odd site (https://github.com/altryne/browserID-jQuery#how-to-use) firstly do a check to see if navigator.id already exists:

```
<script>navigator.id || document.write('<script src="https://login.persona.org/include.js"><\/script>')</script>
```

@6a68 Could we do a similar check at the very top of your script here and quit early if it's true. After that, we don't bother with detecting onload nor loading include.js. Does this look feasible or even ok to do it?

seanmonstar commented:
@chilts its there, at line 54.

chilts commented:
@seanmonstar ah sorry, thanks. :)

jaredhirsch commented:
@chilts this is pretty compact front-end JS, probably not the easiest place to jump in. (Also, doc.write is **_evil**_ ;-)

If you're interested in what's going on here, I'd suggest having a look at the Manning book Third-party JavaScript, which covers a lot of basic techniques for async loading.

jaredhirsch commented:
I'm not sure where this goes from here. Closing the PR because it's just lingering at this point. Might turn this into a self-updating self-hostable snippet some Friday, but going to leave it for now.
